setup.ps1

```diff
@@ -0,0 +1,339 @@
+    }
+    
+    return $selectedIndices
+  }
+  else {
+    # Convert the selection from letter to index and validate
+    if ($userInput -match '^[A-Z]$') {
+      $selectedIndex = [int][char]$userInput - [int][char]'A'
+      if ($selectedIndex -ge 0 -and $selectedIndex -lt $Options.Length) {
+        return $selectedIndex
+      }
+      else {
+        Write-Log "Invalid input: $userInput. Please enter a letter within the valid range of options." -Level 'ERROR'
+        return -1
+      }
+    }
+    else {
+      Write-Log "Invalid input: $userInput. Please enter a letter between A and Z." -Level 'ERROR'
+      return -1
+    }
+  }
+}
+
+function Exit-Script {
+  param (
+    [int]$ExitCode,
+    [bool]$WaitForKeypress = $true
+  )
+
+  if ($Global:OldVirtualPath) {
+    $env:PATH = $Global:OldVirtualPath
+  }
+
+  if ($ExitCode -ne 0) {
+    Write-Log "Script failed. Would you like to open the log file? (Y/N)" -Level 'PROMPT'
+    $openLog = Read-Host
+    if ($openLog -eq 'Y' -or $openLog -eq 'y') {
+      Start-Process notepad.exe -ArgumentList $LogFilePath
+    }
+  }
+  elseif ($WaitForKeypress) {
+    Write-Log "Press any key to exit..."
+    $host.UI.RawUI.ReadKey("NoEcho,IncludeKeyDown") | Out-Null
+  }
+
+  return $ExitCode
+}
+
+# Function to install requirements
+function Install-Requirements {
+  param ([string]$RequirementsPath)
+  
+  if (-not $RequirementsPath) {
+    Write-Log "No requirements path provided for installation." -Level 'ERROR'
+    Exit-Script -ExitCode 1
+  }
+
+  Write-Log "Installing requirements from $RequirementsPath..."
+  $installCmd = if (Test-Path $RequirementsPath) { & $VenvPip install -r $RequirementsPath } else { & $VenvPip install $RequirementsPath }
+  $output = & $installCmd[0] $installCmd[1..$installCmd.Length] 2>&1
+  $output | Out-File $LogFilePath -Append
+  if ($LASTEXITCODE -ne 0) {
+    Write-Log "Conflict detected. Exiting now..." -Level 'ERROR'
+    Exit-Script -ExitCode 1
+  }
+}
+
+function Test-PythonExecutable {
+  param(
+    [string]$PythonExecutable
+  )
+
+  $pythonCmd = Get-Command $PythonExecutable -ErrorAction SilentlyContinue
+  if ($pythonCmd) {
+    $command = "$($pythonCmd.Source) --version 2>&1"
+    $versionOutput = Invoke-Expression $command
+    if ($LASTEXITCODE -eq 0) {
+      $version = $versionOutput | Select-String -Pattern "Python (\d+\.\d+\.\d+)" | ForEach-Object { $_.Matches.Groups[1].Value }
+      Write-Log "Python version $version found using executable '$PythonExecutable'."
+      return $true
+    }
+    else {
+      Write-Log "Python executable '$PythonExecutable' not working correctly." -Level 'ERROR'
+      return $false
+    }
+  }
+  else {
+    Write-Log "Python executable '$PythonExecutable' not found." -Level 'ERROR'
+    return $false
+  }
+}
+
+function Find-PythonExecutable {
+
+  $pythonExecutables = @("python", "python3", "python3.9", "python3.10", "python3.11", "C:\Python39\python.exe", "C:\Python310\python.exe", "C:\Python311\python.exe")
```

Comment:
It would work when the absolute path is correct, but I would need to look through this location too: C:\Users\<User>\AppData\Local\Programs. And add 3.12.